src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -200,99 +206,76 @@ protected TupleBatch fetchNextReady() throws IOException, DbException {
       }
 
       if (record != null) {
+
         /*
-         * Here, we read the first byte of the stream if this is not the first worker. This is simply an optimization.
-         * If the previous character is not a new line, we discard the line.
+         * Here, if we are on the last row, we make sure to read the entire row until we either hit a new line or until
+         * we have read the entire file (this is for the case where a single worker might be reading a single large row
+         * that was split among other workers). If we're at the last row and the last worker is reading, we just mark
+         * this final line as finished.
          */
-        if (lineNumber - 1 == 0 && partitionStartByteRange != 0) {
-          char currentChar =
-              (char)
-                  source.getInputStream(adjustedStartByteRange - 1, adjustedStartByteRange).read();
-          if (currentChar != '\n' && currentChar != '\r') {
-            discardedRecord = true;
+        if (onLastRow && !finishedReadingLastRow && !isLastWorker) {
+          long trailingStartByte = partitionEndByteRange + 1;
+          long trailingEndByte = trailingStartByte + byteOverlap;
+          boolean finalLineFound = false;
+
+          while (!finalLineFound) {
+            boolean withinByteRange = trailingEndByte < maxByteRange;
             /*
-             * This handles closes the stream in corner cases in which a worker might only have a fragment of a tuple as
-             * its entire partition. Since the server makes sure to provide workers a partition with a minimum size,
-             * this case could only happen with very large rows.
+             * If we are within the max byte range, then keep checking for a new line. Otherwise, if we've reached the
+             * end of the file, mark finalLineFound as true.
              */
-            if (onLastRow) {
-              parser.close();
-              break;
+            if (withinByteRange) {
+              InputStream trailingEndInputStream =
+                  source.getInputStream(trailingStartByte, trailingEndByte);
+              int dataChar = trailingEndInputStream.read();
+              while (dataChar != -1) {
+                char currentChar = (char) dataChar;
+                if (currentChar == '\n' || currentChar == '\r') {
+                  finalLineFound = true;
+                  break;
+                }
+                dataChar = trailingEndInputStream.read();
+              }
+              trailingEndInputStream.close();
```

Comment:
Can't we remember the position of the newline byte before we close the stream so we can use it to construct the next `InputStream`? Since we're reading bytes, not characters, I think this should be unproblematic.